File: sudoku.c
#include<stdio.h>
#include<stdlib.h>
#include "sudoku_struct.h"
int _sudoku_solve = 0;

int output_to_file(FILE* file, Sudoku s){
    for(int i = 0; i < 9; i++){
        for(int j = 0; j < 9; j++){
            fprintf(file, j == 8 ? "%d\n": "%d ", s.sudoku[i][j]);
        }
    }
    return 0;
}

int load_from_file(FILE* file, Sudoku * s){
    char tmp[20];
    int cnt = 0;
    for(int line = 0; line < 9; line++){
        for(int num = 0; num < 9; num++){
            int tmp;
            if(fscanf(file, "%d", &tmp) !=1) return 1;
            if(tmp > 9 || tmp < 0) return 1;
            if(tmp) cnt++;
            s->sudoku[line][num] = tmp;
        }
    }
    s->filled = cnt;
    return 0;
}
int check_sudoku(Sudoku* s){
    // check line
    for(int i = 0; i < 9; i++){
        char check[10] = {0};
        for(int j = 0; j < 9; j++){
            if(s->sudoku[i][j] == 0) continue;
            if(check[s->sudoku[i][j]]) return 0;
            else check[s->sudoku[i][j]] = 1;
        }
    }
    // check column
    for(int j = 0; j < 9; j++){
        char check[10] = {0};
        for(int i = 0; i < 9; i++){
            if(s->sudoku[i][j] == 0) continue;
            if(check[s->sudoku[i][j]]) return 0;
            else check[s->sudoku[i][j]] = 1;
        }
    }
    for(int block_i = 0; block_i < 3; block_i++) for(int block_j = 0; block_j < 3; block_j++){
        char check[9] = {0};
        for(int i = 0; i < 3; i++) for(int j = 0; j < 3; j++){
            if(s->sudoku[block_i*3 + i][block_j*3 + j] == 0) continue;
            if(check[s->sudoku[block_i*3+i][block_j*3+j]]) return 0;
            else check[s->sudoku[block_i*3+i][block_j*3+j]] = 1;
        }
    }
    return 1;
}

Sudoku generate_sudoku(){
    int first_num = 3; // (2+4) % 9 +1
    int init_matrix[][9][9]={
      { { 1, 2, 3, 4, 5, 6, 7, 8, 9 },
        { 4, 5, 6, 7, 8, 9, 1, 2, 3 },
        { 7, 8, 9, 1, 2, 3, 4, 5, 6 },
        { 2, 1, 4, 3, 6, 5, 8, 9, 7 },
        { 3, 6, 5, 8, 9, 7, 2, 1, 4 },
        { 8, 9, 7, 2, 1, 4, 3, 6, 5 },
        { 5, 3, 1, 6, 4, 2, 9, 7, 8 },
        { 6, 4, 2, 9, 7, 8, 5, 3, 1 },
        { 9, 7, 8, 5, 3, 1, 6, 4, 2 } },
      { { 3, 9, 4, 5, 1, 7, 6, 2, 8 }, 
        { 5, 1, 7, 6, 2, 8, 3, 9, 4 },
        { 6, 2, 8, 3, 9, 4, 5, 1, 7 },
        { 9, 3, 5, 4, 7, 1, 2, 8, 6 },
        { 4, 7, 1, 2, 8, 6, 9, 3, 5 },
        { 2, 8, 6, 9, 3, 5, 4, 7, 1 },
        { 1, 4, 3, 7, 5, 9, 8, 6, 2 },
        { 7, 5, 9, 8, 6, 2, 1, 4, 3 },
        { 8, 6, 2, 1, 4, 3, 7, 5, 9 } },
      { { 7, 6, 1, 9, 8, 4, 2, 3, 5 }, 
        { 9, 8, 4, 2, 3, 5, 7, 6, 1 },
        { 2, 3, 5, 7, 6, 1, 9, 8, 4 },
        { 6, 7, 9, 1, 4, 8, 3, 5, 2 },
        { 1, 4, 8, 3, 5, 2, 6, 7, 9 },
        { 3, 5, 2, 6, 7, 9, 1, 4, 8 },
        { 8, 1, 7, 4, 9, 6, 5, 2, 3 },
        { 4, 9, 6, 5, 2, 3, 8, 1, 7 },
        { 5, 2, 3, 8, 1, 7, 4, 9, 6 } },
      { { 7, 1, 5, 4, 3, 6, 2, 9, 8 }, 
        { 4, 3, 6, 2, 9, 8, 7, 1, 5 },
        { 2, 9, 8, 7, 1, 5, 4, 3, 6 },
        { 1, 7, 4, 5, 6, 3, 9, 8, 2 },
        { 5, 6, 3, 9, 8, 2, 1, 7, 4 },
        { 9, 8, 2, 1, 7, 4, 5, 6, 3 },
        { 3, 5, 7, 6, 4, 1, 8, 2, 9 },
        { 6, 4, 1, 8, 2, 9, 3, 5, 7 },
        { 8, 2, 9, 3, 5, 7, 6, 4, 1 } } };
    int mat_num = rand() % 4;
    Sudoku rt;
    // init
    for(int i = 0; i < 9; i++) for(int j = 0; j < 9; j++)
        rt.sudoku[i][j] = init_matrix[mat_num][i][j];
    // shift line
    for(int i = 0; i < 20; i++){
        int block_num = rand() % 3;
        int first_shift = rand() % 3;
        int second_shift = rand() % 2;
        if(second_shift >= first_shift) second_shift ++;
        first_shift += block_num * 3;
        second_shift += block_num * 3;
        int tmp;
        for(int i = 0; i < 9; i++){
            tmp = rt.sudoku[first_shift][i];
            rt.sudoku[first_shift][i] = rt.sudoku[second_shift][i];
            rt.sudoku[second_shift][i] = tmp;
        }
    }
    //shift column
    for(int i = 0; i < 20; i++){
        int block_num = rand() % 3;
        int first_shift = rand() % 3;
        int second_shift = rand() % 2;
        if(second_shift >= first_shift) second_shift ++;
        first_shift += block_num * 3;
        second_shift += block_num * 3;
        int tmp;
        for(int i = 0; i < 9; i++){
            tmp = rt.sudoku[i][first_shift];
            rt.sudoku[i][first_shift] = rt.sudoku[i][second_shift];
            rt.sudoku[i][second_shift] = tmp;
        }
    }
    if(rt.sudoku[0][0] != first_num){
        int swap_1 = first_num, swap_2 = rt.sudoku[0][0];
        for(int i = 0; i < 9; i++) for(int j = 0; j < 9; j++){
            if(rt.sudoku[i][j] == swap_1) rt.sudoku[i][j] = swap_2;
            else if(rt.sudoku[i][j] == swap_2) rt.sudoku[i][j] = swap_1;
        }
    }
    if(check_sudoku(&rt)) puts("GG");
    return rt;
}

int check_one(Sudoku * s, int i, int j, int num){
    int flag = 1;
    for(int _i = 0; _i < 9; _i++)
        if(s->sudoku[_i][j] == num) flag = 0;
    for(int _j = 0; _j < 9; _j++)
        if(s->sudoku[i][_j] == num) flag = 0;
    int block_i = i / 3 * 3, block_j = j / 3 * 3;
    for(int _i = block_i; _i < block_i + 3; _i++)
        for(int _j = block_j; _j < block_j + 3; _j++)
            if(s->sudoku[_i][_j] == num) flag = 0;
    return flag;
}

int max_fill(Sudoku * s, int x[], int y[]){
    int fill_cnt = 0;
    int flag = 0;
    while(s->filled < 81){
        //fill line
        flag = 0;
        for(int i = 0; i < 9; i++){
            int sum = 0, cnt = 0, emp_pos;
            for(int j = 0; j < 9; j++){
                if(s->sudoku[i][j]){
                    sum += s->sudoku[i][j];
                    cnt ++;
                }
                else emp_pos = j;
            }
            if(cnt == 8){
                int to_fill = 45 - sum;
                if(!check_one(s, i, emp_pos, to_fill)) return -fill_cnt;
                s->sudoku[i][emp_pos] = to_fill;
                flag = 1;
                x[fill_cnt] = i;
                y[fill_cnt] = emp_pos;
                fill_cnt++;
                s->filled++;
            }
        }
        if(flag) break;
        for(int j = 0; j < 9; j++){
            int sum = 0, cnt = 0, emp_pos;
            for(int i = 0; i < 9; i++){
                if(s->sudoku[i][j]){
                    sum += s->sudoku[i][j];
                    cnt ++;
                }
                else emp_pos = i;
            }
            if(cnt == 8){
                int to_fill=  45 - sum;
                if(!check_one(s, emp_pos, j, to_fill)) return -fill_cnt; 
                s->sudoku[emp_pos][j] = to_fill;
                flag = 1;
                x[fill_cnt] = emp_pos;
                y[fill_cnt] = j;
                fill_cnt++;
                s->filled++;
            }
        }
        if(flag) break;
        for(int block_i = 0; block_i < 3; block_i++) for(int block_j = 0; block_j < 3; block_j++){
            int sum = 0, cnt = 0, emp_pos_x, emp_pos_y;
            for(int i = block_i*3; i < block_i*3 +3; i++) for(int j = block_j*3;j < block_j*3 + 3; j++){
                if(s->sudoku[i][j]){
                    sum += s->sudoku[i][j];
                    cnt++;
                }
                else{
                    emp_pos_x = i;
                    emp_pos_y = j;
                }
            }
            if(cnt == 8){
                int to_fill = 45 - sum;
                if(!check_one(s, emp_pos_x, emp_pos_y, to_fill)) return -fill_cnt;
                s->sudoku[emp_pos_x][emp_pos_y] = to_fill;
                flag = 1;
                x[fill_cnt] = emp_pos_x;
                y[fill_cnt] = emp_pos_y;
                fill_cnt++;
                s->filled++;
            }
        }
        if(!flag) break;
    }
    return fill_cnt;
}

void print_sudoku(Sudoku * s){
    for(int i = 0; i < 9; i ++){
        for(int j = 0; j < 9; j++)
            printf(j == 8 ? "%d\n" : "%d ", s->sudoku[i][j]);
    }
}

void _find_empty(Sudoku * s, int * i, int * j){
    for(*i = 0; *i < 9; (*i)++) for(*j = 0; *j < 9; (*j)++)
        if(s->sudoku[*i][*j] == 0) return;
}

int error = 0;
void solve_traceback(Sudoku * s){
    if(_sudoku_solve) return;
    int _cnt = 0;
    for(int i = 0; i < 9; i++) for(int j = 0; j < 9; j++)
        if(s->sudoku[i][j]) _cnt ++;
    if(_cnt != s->filled) error=1;
    if(s->filled == 81){
        _sudoku_solve = 1;
        print_sudoku(s);
        return;
    }
    //print_sudoku(s);
    //printf("solve: %d\n", s->filled);
    //putchar('\n');
    //int fill_x[30], fill_y[30];
    //int fill_num = max_fill(s, fill_x, fill_y);
    //if(fill_num < 0){
    //    //traceback
    //    s->filled += fill_num;
    //    for(int i = 0; i < fill_num; i++)
    //        s->sudoku[fill_x[i]][fill_y[i]] = 0;
    //    return;
    //}
    //s->filled += fill_num;
    //print_sudoku(s);
    //printf("solve: %d\n", s->filled);
    //putchar('\n');
    if(s->filled == 81){
        _sudoku_solve = 1;
        return;
    }
    //find the empty point
    int i, j = 0;
    _find_empty(s, &i, &j);
    s->filled++;
    for(int num = 1; num < 10; num++){
        int flag = check_one(s, i, j, num);
        if(flag){
            s->sudoku[i][j] = num;
            solve_traceback(s);
        }
    }
    //traceback
    s->filled--;
    //s->filled -= fill_num;
    s->sudoku[i][j] = 0;
    //for(i = 0; i < fill_num; i++)
    //    s->sudoku[fill_x[i]][fill_y[i]] = 0;
}

File: sudoku.py
import numpy as np
import random
import itertools

class Sudoku:

    __solved = False

    init_matrixs = np.array([
      [ [ 1, 2, 3, 4, 5, 6, 7, 8, 9 ],
        [ 4, 5, 6, 7, 8, 9, 1, 2, 3 ],
        [ 7, 8, 9, 1, 2, 3, 4, 5, 6 ],
        [ 2, 1, 4, 3, 6, 5, 8, 9, 7 ],
        [ 3, 6, 5, 8, 9, 7, 2, 1, 4 ],
        [ 8, 9, 7, 2, 1, 4, 3, 6, 5 ],
        [ 5, 3, 1, 6, 4, 2, 9, 7, 8 ],
        [ 6, 4, 2, 9, 7, 8, 5, 3, 1 ],
        [ 9, 7, 8, 5, 3, 1, 6, 4, 2 ] ],
      [ [ 3, 9, 4, 5, 1, 7, 6, 2, 8 ], 
        [ 5, 1, 7, 6, 2, 8, 3, 9, 4 ],
        [ 6, 2, 8, 3, 9, 4, 5, 1, 7 ],
        [ 9, 3, 5, 4, 7, 1, 2, 8, 6 ],
        [ 4, 7, 1, 2, 8, 6, 9, 3, 5 ],
        [ 2, 8, 6, 9, 3, 5, 4, 7, 1 ],
        [ 1, 4, 3, 7, 5, 9, 8, 6, 2 ],
        [ 7, 5, 9, 8, 6, 2, 1, 4, 3 ],
        [ 8, 6, 2, 1, 4, 3, 7, 5, 9 ] ],
      [ [ 7, 6, 1, 9, 8, 4, 2, 3, 5 ], 
        [ 9, 8, 4, 2, 3, 5, 7, 6, 1 ],
        [ 2, 3, 5, 7, 6, 1, 9, 8, 4 ],
        [ 6, 7, 9, 1, 4, 8, 3, 5, 2 ],
        [ 1, 4, 8, 3, 5, 2, 6, 7, 9 ],
        [ 3, 5, 2, 6, 7, 9, 1, 4, 8 ],
        [ 8, 1, 7, 4, 9, 6, 5, 2, 3 ],
        [ 4, 9, 6, 5, 2, 3, 8, 1, 7 ],
        [ 5, 2, 3, 8, 1, 7, 4, 9, 6 ] ],
      [ [ 7, 1, 5, 4, 3, 6, 2, 9, 8 ], 
        [ 4, 3, 6, 2, 9, 8, 7, 1, 5 ],
        [ 2, 9, 8, 7, 1, 5, 4, 3, 6 ],
        [ 1, 7, 4, 5, 6, 3, 9, 8, 2 ],
        [ 5, 6, 3, 9, 8, 2, 1, 7, 4 ],
        [ 9, 8, 2, 1, 7, 4, 5, 6, 3 ],
        [ 3, 5, 7, 6, 4, 1, 8, 2, 9 ],
        [ 6, 4, 1, 8, 2, 9, 3, 5, 7 ],
        [ 8, 2, 9, 3, 5, 7, 6, 4, 1 ] ] ], dtype=np.int8)

    def generate_random_sudoku(self):
        first_num = 3 #(2+4) % 9 + 1
        self.sudoku = self.init_matrixs[random.randrange(0,4)]
        for _ in range(20):
            # 行变换
            block_num = random.randrange(0,3)
            p, q = random.choices([i for i in range(block_num*3, block_num*3+3)], k=2)
            self.sudoku[[p,q],:] = self.sudoku[[q,p],:]
        for _ in range(20):
            # 行变换
            block_num = random.randrange(0,3)
            p, q = random.choices([i for i in range(block_num*3, block_num*3+3)], k=2)
            self.sudoku[:,[p,q]] = self.sudoku[:,[q,p]]
        if self.sudoku[0][0] == first_num:
            return
        cur_num = self.sudoku[0][0]
        for i, j in itertools.product(range(9), repeat=2):
            if self.sudoku[i][j] == cur_num:
                self.sudoku[i][j] = first_num
            elif self.sudoku[i][j] == first_num:
                self.sudoku[i][j] = cur_num

    def check(self):
        # row check
        for row in self.sudoku:
            _check = [False] * 10
            for num in row:
                if num == 0: continue
                if _check[num]: return False
                else: _check[num] = True
        # column check
        for column in self.sudoku.T:
            _check = [False] * 10
            for num in column:
                if num == 0: continue
                if _check[num]: return False
                else: _check[num] = True
        # 3*3 check
        for i,j in itertools.product((0,1,2), repeat=2):
            matrix_s = self.sudoku[i*3:i*3+3, j*3:j*3+3].reshape((9))
            _check = [False] * 10
            for num in matrix_s:
                if num == 0: continue
                if _check[num]: return False
                else: _check[num] = True
        return True

    def __init__(self, s=None):
        if not s:
            self.generate_random_sudoku()            
            assert self.check()
        elif isinstance(s, Sudoku):
            self.sudoku = s.sudoku.copy()
        elif isinstance(s, str):
            # self.import_check(str)
            s = s.strip()
            s = s.split('\n')
            self.sudoku = np.array([list(map(lambda x: int(x), i.split())) for i in s])
            #self.sudoku = np.array(list(map(lambda x: int(x),map(lambda x: x.split(), s))))
            # add check
        else:
            raise Exception
    
    def __str__(self):
        lines = []
        for l in self.sudoku:
            lines.append(' '.join(map(lambda x: str(x), l.tolist())))
        return '\n'.join(lines)

    def max_fill(self):
        while True:
            if np.count_nonzero(self.sudoku) == 81: 
                return
            flag = False
            # fill row
            for i in range(9):
                if np.count_nonzero(self.sudoku[i]) == 8:
                    flag = True
                    to_fill_num = 45 - self.sudoku[i].sum()
                    for j in range(9):
                        if self.sudoku[i][j] == 0:
                            self.sudoku[i][j] = to_fill_num
                            break
            # fill collomn
            for j in range(9):
                if np.count_nonzero(self.sudoku[:,j]) == 8:
                    flag = True
                    to_fill_num = 45 - self.sudoku[:,j].sum()
                    for i in range(9):
                        if self.sudoku[i][j] == 0:
                            self.sudoku[i][j] = to_fill_num
                            break
            # fill 3*3
            for i,j in itertools.product((0,1,2), repeat=2):
                if np.count_nonzero(self.sudoku[3*i:3*i+3, 3*j:3*j+3]) == 8:
                    flag = True
                    to_fill_num = 45 - self.sudoku[3*i:3*i+3, 3*j:3*j+3].sum()
                    for p, q in itertools.product(range(3*i, 3*i+3), range(3*j, 3*j+3)):
                        if self.sudoku[p][q] == 0:
                            self.sudoku[p][q] = to_fill_num
                            break
            if not flag: break 

    def solve(self):
        type(self).__solved = False
        self.__traceback()
        if not type(self).__solved:
            print("no solution")
        else:
            print(str(self))

    def __traceback(self):
        print(str(self))
        print('')
        if type(self).__solved:
            return
        if np.count_nonzero(self.sudoku) == 81:
            type(self).__solved = True
            return
        # search the unfill
        self.max_fill()
        for i,j in itertools.product([_ for _ in range(9)], repeat=2):
            if self.sudoku[i][j] != 0:
                break
        candidate = [True] * 10
        for _j in range(9):
            candidate[self.sudoku[i][_j]] = False
        for _i in range(9):
            candidate[self.sudoku[_i][j]] = False
        block_i = i // 3 * 3
        block_j = j // 3 * 3
        for _i, _j in itertools.product(range(block_i, block_i + 3), range(block_j, block_j + 3)):
            candidate[self.sudoku[_i][_j]] = False
        for _ in range(1, 10):
            if candidate[_]:
                next_suduku = Sudoku(self)
                next_suduku.sudoku[i][j] = _
                next_suduku.__traceback()
            if type(self).__solved:
                return
        
    def import_check(self, string):
        lines = string.splitlines(False)
        if len(lines) != 9:
            raise Exception("format error")
        for line in lines:
            nums = line.split()
            if len(nums) != 9:
                raise Exception("format error")
            for num in nums:
                if int(num) < 0 or int(num) > 9:
                    raise Exception("format error")

if __name__ == "__main__":
    with open('./sudoku.txt', 'r') as f:
        data = f.read()
    a = Sudoku(data)
    a.max_fill()
    print(a)